### PR TITLE
:zap: adds ability to run the dashboard from any mount point

### PR DESCRIPTION
**Pull request by flovilmart**
- fixes allowInsecureHTTP

### CODE REVIEW

File: Parse-Dashboard/app.js [hunk 1]
```diff
@@ -83,7 +83,28 @@ module.exports = function(config) {
 
   // For every other request, go to index.html. Let client-side handle the rest.
   app.get('/*', function(req, res) {
-    res.sendFile(__dirname + '/index.html');
+    let url = req.url;
```
Comment (flovilmart) [inline Parse-Dashboard/app.js]: infer the mount path

File: Parse-Dashboard/app.js [hunk 2]
```diff
@@ -83,7 +83,28 @@ module.exports = function(config) {
...
+    var mountPathLength = req.originalUrl.length - req.url.length;
+    var mountPath = req.originalUrl.slice(0, mountPathLength);
+    if (!mountPath.endsWith('/')) {
+      mountPath += '/';
+    }
+    res.send(`<!DOCTYPE html>
+      <head>
+        <base href="${mountPath}"/>
+        <script>
+          PARSE_DASHBOARD_CONFIG_URI = "${mountPath}parse-dashboard-config.json";
```
Comment (flovilmart) [inline Parse-Dashboard/app.js]: instead of passing the mount, just pass the configURI, that's cleaner

File: src/components/Icon/Icon.react.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ let Icon = ({ name, fill, width, height }) => {
   }
   return (
     <svg {...props} >
-      <use xlinkHref={`/bundles/sprites.svg#${name}`} />
+      <use xlinkHref={`bundles/sprites.svg#${name}`} />
```
Comment (flovilmart) [inline src/components/Icon/Icon.react.js]: no / as it should be relative to base

File: Parse-Dashboard/app.js [hunk 4]
```diff
@@ -83,7 +83,28 @@ module.exports = function(config) {
...
+    let url = req.url;
+    let originalUrl = req.originalUrl;
+    var mountPathLength = req.originalUrl.length - req.url.length;
+    var mountPath = req.originalUrl.slice(0, mountPathLength);
+    if (!mountPath.endsWith('/')) {
+      mountPath += '/';
+    }
+    res.send(`<!DOCTYPE html>
+      <head>
+        <base href="${mountPath}"/>
```
Comment (flovilmart) [inline Parse-Dashboard/app.js]: needed to properly route the request

File: src/dashboard/Dashboard.js [hunk 5]
```diff
@@ -117,7 +116,7 @@ class Dashboard extends React.Component {
   }
 
   componentDidMount() {
-    get('/parse-dashboard-config.json').then(({ apps, newFeaturesInLatestVersion = [] }) => {
+    get(this.props.configURI).then(({ apps, newFeaturesInLatestVersion = [] }) => {
```
Comment (flovilmart) [inline src/dashboard/Dashboard.js]: get the config URI

File: src/dashboard/index.js [hunk 6]
```diff
@@ -17,4 +17,5 @@ import 'babel-polyfill';
 require('stylesheets/fonts.scss');
 installDevTools(Immutable);
 
-ReactDOM.render(<Dashboard/>, document.getElementById('browser_mount'));
+var configURI = window.PARSE_DASHBOARD_CONFIG_URI || '/parse-dashboard-config.json';
+ReactDOM.render(<Dashboard configURI={configURI}/>, document.getElementById('browser_mount'));
```
Comment (flovilmart) [inline src/dashboard/index.js]: pass it as a prop, that kinda makes sense as one would be able to mount multiple dash with multiple auth

**Comment by mamaso** [comment]
Good stuff, definitely less intrusive. I like how you inferred the mount path.

 A few issues that you might want to test:
- does it load the phone images on the push tab?
- are there any failing requests in the network tab of developer tools?

**Comment by flovilmart** [comment]
You're free to test!

**Comment by drew-gross** [comment]
Turns out the images in the push page don't work. Otherwise this looks sweet though.

**Comment by flovilmart** [comment]
Yep, also the internal API, need to inject dashboardURL into the apps.

**Comment by drew-gross** [comment]
How would you feel about just injecting the config into the page directly? Then an extra fetch would not be necessary. Also, if you are moving the index.html into the source, can you also delete the actual index.html file?

**Comment by flovilmart** [comment]
Yeah I was thinking about that too.
We can't really remove the endpoint as we use it for authentication purposes and that would expose the keys, but we could rename it to /config.json

**Comment by drew-gross** [comment]
Ah, that is unfortuanate :(

**Comment by flovilmart** [comment]
no I'm getting through it!

**Comment by drew-gross** [comment]
Just noticed the favicon is missing too. Obviously that is lower-pri :p

**Comment by flovilmart** [comment]
Just remains the internal API calls in lib/ParseApp.js

**Comment by flovilmart** [comment]
@drew-gross that should be it!

File: src/lib/AJAX.js [hunk 7]
```diff
@@ -9,8 +9,22 @@ import * as CSRFManager from 'lib/CSRFManager';
...
+export function setBasePath(newBasePath) {
+  basePath = newBasePath || '';
+  if (basePath.endsWith('/')) {
+    basePath = basePath.slice(0, basePath.length-1);
+  }
+}
+
 // abortable flag used to pass xhr reference so user can abort accordingly
 export function request(method, url, body, abortable = false, withCredentials = true, useRequestedWith = true) {
+  if (!url.startsWith('http://')
```
Comment (flovilmart) [inline src/lib/AJAX.js]: open to suggestions about that, that's the less invasive I've found

Comment (drew-gross) [inline src/lib/AJAX.js]: Yeah this is fine. The only things that aren't going to go through the Parse sdk are the config loading, and potentially this app creation stuff that is being discussed.

**Comment by fabiogomezdiaz** [comment]
@flovilmart great job with this PR! With this, will we now be allowed to deploy the dashboard to Heroku? If so, how could I go about that? Do I need to set any specific environment/config variables?

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [4.0.0-alpha.2](https://github.com/parse-community/parse-dashboard/releases/tag/4.0.0-alpha.2)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [4.0.0-beta.1](https://github.com/parse-community/parse-dashboard/releases/tag/4.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [4.0.0](https://github.com/parse-community/parse-dashboard/releases/tag/4.0.0)

**Pull request merged**